exonum/src/runtime/mod.rs

```diff
@@ -41,19 +41,40 @@
 //!   What deployment entails depends on the runtime; e.g., the artifact may be downloaded
 //!   by each Exonum node, verified for integrity and then added into the execution environment.
 //!
-//! 3. Once the artifact is deployed, it is possible to instantiate a corresponding service.
+//! 3. For each node, an artifact may be deployed either asynchronously or synchronously /
+//!   in a blocking manner. The supervisor usually first commands a node to deploy the artifact
+//!   asynchronously via [`Mailbox`] once the decision to start deployment is reached
+//!   by the blockchain administrators.
+//!
+//! 4. Async deployment usually has a deadline and/or success and failure conditions. As an example,
+//!   the supervisor may collect confirmations from the validator nodes that have
+//!   successfully deployed the artifact, and once all the validator nodes have sent
+//!   their confirmations, the artifact is *committed*.
```

Comment:
As there is no 'undeploy' operation in case when some nodes failed to deploy the artifact, is that correct that the runtimes do not know if and when the artifact is actually committed and just rely on the (?) dispatcher issuing `start_adding_service` only if the conditions are met?

What would happen if
1. Most nodes deployed an artifact correctly, but a couple of nodes failed (e.g., the admins forgot to put the artifact where it belongs). As a result, the runtimes on the 'good' nodes have it deployed; on the 'bad' — don't. The runtimes on the 'good' don't know that it was not committed.

2. The administrators attempt a re-deploy. This time, as the 'good' nodes already have this artifact, the operation fails there (just because of an illegal attempt to deploy it twice); and succeeds on the 'bad' ones.

Reply:
So, "consensus-agreed result" = failure -> deploy of this particular artifact failed in the whole network, but 'good' runtimes will _retain_ the artifact as deployed till _restarted_ -> the redeploy of the same artifact is impossible without restart, is that correct?

I think a deploy error on some but all nodes is a realistic use-case (but the one we can probably agree to ignore in the upcoming release). Would a 'Runtime#unloadArtifact' operation, that the dispatcher invokes when the operation failed in terms of network but completed successfully on this node (i.e., if I remember the current policy correctly, failed on _any_ single node), help? Or `Runtime#unloadIfPresent` if we'd like the dispatcher to not have local data on whether the local deployment was successful — but it might already have that.

Reply:
> the redeploy of the same artifact is impossible without restart

@slowli , @aleksuss Is this limitation planned to be addressed?

Reply:
AFAIK yes, we plan it.